runtime/http_client.go

```diff
@@ -21,32 +21,68 @@
 package zanzibar
 
 import "net/http"
-import "go.uber.org/zap"
+import (
+	"net"
+	"time"
+
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+)
 
 // HTTPClient defines a http client.
 type HTTPClient struct {
 	gateway *Gateway
 
 	Client  *http.Client
 	Logger  *zap.Logger
+	Scope   *tally.Scope
 	BaseURL string
 }
 
 // NewHTTPClient will allocate a http client.
 func NewHTTPClient(
 	gateway *Gateway, baseURL string,
 ) *HTTPClient {
+	scope := gateway.MetricScope.SubScope("") // TODO: get http client name
+	timer := scope.Timer("dial")
+	success := scope.Counter("dial.success")
+	error := scope.Counter("dial.error")
+
 	return &HTTPClient{
 		gateway: gateway,
-
-		Logger: gateway.Logger,
+		Logger:  gateway.Logger,
 		Client: &http.Client{
 			Transport: &http.Transport{
 				DisableKeepAlives:   false,
 				MaxIdleConns:        500,
 				MaxIdleConnsPerHost: 500,
+				Dial:                getInstrumentedDial(timer, success, error),
 			},
 		},
 		BaseURL: baseURL,
 	}
 }
+
+func getInstrumentedDial(timer tally.Timer, successCtr, errorCtr tally.Counter) func(string, string) (net.Conn, error) {
+	return func(n, a string) (conn net.Conn, err error) {
+		stat := instrument(timer, successCtr, errorCtr)
+		defer func() {
```

Comment:
nitpick: i find defers hard to read because of unwinding.

```
stat := instrument(timer, successCtr, errorCtr);
conn, err := net.Dial(n, a)
stat(err)
return conn, err
```

is easier to read.